### ISSUE TITLE
Support for time.sleep()

### ISSUE THREAD

**Issue by dschreij** [QQuick/Transcrypt#537]
I know this is a difficult one to implement due to JavaScript's asynchronous nature, but do you think there will be support for the `time.sleep()` function which pauses execution for the passed number of seconds?

I understand this is dodgy, as under the hood something like this has to be implemented with a callback or promise, but now async/await has arrived to JS, it might be (conceptually) easier to realize?

**Comment by the1337guy** [comment]
that'd force every function that uses time.sleep to be async, which will make every other function that uses the said function to be async, and so on. You can't serially get your results from Promise-returning functions

**Comment by theodox** [comment]
Is this because you actually want to block _everything_ for some length of time?

**Comment by dschreij** [comment]
Yes, sort of. We are translating a psychologic [experiment builder](http://osdoc.cogsci.nl/) (meant as reference and not as a shameless plug ;) ) from Python to Javascript so the experiments created in the builder can be run online. The builder helps with create stimulus displays for the experiment. People can write script for small pieces of the experiment if they want, and thus can do something like this:

~~~python
# Create some buffers/canvases
c1 = canvas()
c2 = canvas()
c3 = canvas()

# Draw one shape on each buffer
c1.square(x=100, y=100, w=10, h=10)
c2.circle(x=100, y=100, r=10)
c3.text(x=100, y=100, "Some text")

# Show each display and pause for 1 second in between
c1.show()
time.sleep(1)
c2.show()
time.sleep(1)
c3.show()
~~~

So this kind of thing is very hard to convert to JS, because you can't pause executing, unless you start working with promises or async/await, and even such solutions are pretty artificial. And, as @the1337guy states, all functions from the top on need to be declared as async, which I agree is quite impractical.

**Comment by theodox** [comment]
Unless it's important to completely block the entire execution, something like this might work:

```
def timer(length):
    def timer_elapse(resolve): 
        setTimeout(lambda: resolve(), length * 1000)

    return __new__(Promise(timer_elapse, lambda: console.log("exception in timer")))


async def sequence(*functions, delay = 1.0):
    for each_fn in functions:
        await(timer(delay))
        each_fn()


def test_1():
    print ("did one")

def test_2():
    print ("did two")
sequence(test_1, test_2)
```

Only the sequencing function needs to be declared async, the others can be regular functions

**Comment by dschreij** [comment]
Thanks for the example! This requires people to convert their python scripts from using time.sleep() to the structure you propose here, I suppose? We'll take it into consideration as a possible approach. Ideally we'd want people's scripts to work without any adapations (so translate the Python code 'as-is' to JavaScript), but I am coming more and more to the conclusion that this currently is a bridge too far.

**Comment by JdeH** [comment]
Sleep () and also blocking I/O are very hard and costly to solve in a browser in terms of complexity, generated code size and performance.
Browser manufacturers don't like "sleep ()" and blocking I/O because it can make users wait.
The browser platform is by nature event driven, so not very suitable for terminal-like, conversational mode behaviour, unless such behaviour stems from the browser merely being a window on a "classical" server.

Implementing "time.sleep ()" requires a totally different approach to compilation, involving something called "continuations".
As far as I know there's one tool that took that route, namely Skulpt.


http://www.skulpt.org


Its prime target is education, for which it is a great tool.
The prime target of Transcrypt is offering JS-like performance, page load times and code size in a browser environment, as well as offering seamless access to existing JS libraries (which mostly revolve around event driven GUI's.)

**Comment by dschreij** [comment]
Thanks for the explanation. I've discovered skulpt a while ago and it looked impressive too, but it appears to no longer be maintained from the looks of its git records. I am a JS programmer myself and I understand what you're saying. The event-loop based implementation is just a completely different philosophy from the linear execution of Python or most other programming languages.
I'm happy with the async/await structure addition to JS, but I see that it doesn't help that much in making a sleep function for Transcypt possible. Keep up the good work nevertheless!

**Comment by dschreij** [comment]
Feel free to close this issue by the way
